### PR TITLE
Early termination in back prop

### PR DESCRIPTION
**Pull request by timmoon10**
~~This PR builds on #220, so that should be accepted first.~~

Major changes:
- Back prop stops early if all required gradients have already been computed.
- We now handle the case where weights are being shared and optimizers use non-blocking allreduces.
- CPU layers keep all activations and error signals on CPU and GPU layers on GPU.

I get a ~10% performance boost since we no longer have to do a CPU-GPU data transfer during backprop in the input layer.

### CODE REVIEW

File: src/layers/layer.cpp [hunk 1]
```diff
@@ -247,99 +247,28 @@ void Layer::forward_prop() {
   // Setup matrix data, e.g. input matrices
   fp_setup_data(m_model->get_current_mini_batch_size());
 
-#ifdef LBANN_HAS_CUDNN
-  if(m_using_gpus) {
```
Comment (timmoon10) [inline src/layers/layer.cpp]: GPU matrices for previous activations are now setup in `fp_setup_data()`.

File: src/layers/layer.cpp [hunk 2]
```diff
@@ -247,99 +247,28 @@ void Layer::forward_prop() {
...
-                               m_num_neurons :
-                               m_activations_d[0].get_leading_dim());
-    cudnn::set_tensor_cudnn_desc(m_prev_activations_cudnn_desc,
-                                 m_mini_batch_size_per_gpu,
-                                 m_prev_neuron_dims,
-                                 input_stride);
-    cudnn::set_tensor_cudnn_desc(m_activations_cudnn_desc,
-                                 m_mini_batch_size_per_gpu,
-                                 m_neuron_dims,
-                                 output_stride);
```
Comment (timmoon10) [inline src/layers/layer.cpp]: These descriptors are now initialized in `fp_setup_data()`.

File: src/layers/layer.cpp [hunk 3]
```diff
@@ -1118,10 +1093,42 @@ void Layer::get_gpu_fp_output(cudnn::matrix& output_d,
...
+      El::Copy(activation, workspace);
+    }
+
+    // Initialize GPU matrix
+    auto&& cudnn = output_d.get_cudnn_manager();
+    const int num_gpus = cudnn->get_num_gpus();
+    const int num_processes = m_comm->get_procs_per_model();
+    const int max_local_mini_batch_size = (m_model->get_max_mini_batch_size() + num_processes - 1) / num_processes;
+    const int max_mini_batch_size_per_gpu = (max_local_mini_batch_size + num_gpus - 1) / num_gpus;
+    output_d.resize(get_num_neurons(child_index), max_mini_batch_size_per_gpu);
```
Comment (timmoon10) [inline src/layers/layer.cpp]: This should only allocate memory at the first iteration.

File: src/optimizers/optimizer.cpp [hunk 4]
```diff
@@ -163,62 +139,56 @@ AbsDistMat& optimizer::get_gradient() {
...
-      m_comm->allreduce(*m_staging, m_staging->RedundantComm());
-    }
-#else
-    m_comm->allreduce(*m_staging, m_staging->RedundantComm());
-#endif  // LBANN_NBALLREDUCE_GRADIENT
-    add_to_gradient(*m_staging);
-    El::Zero(*m_staging);
-    m_cpu_staging_is_nonzero = false;
+  // Check if all gradient sources have made contributions
+  m_gradient_sources.erase(nullptr);
```
Comment (ndryden) [inline src/optimizers/optimizer.cpp]: What's with erasing `nullptr`?

Comment (timmoon10) [reply src/optimizers/optimizer.cpp]: Paranoia mostly.

**Pull request merged**